require namespace before VRG creation #240

tjanssen3 opened the pull request:
- if a workload is used for a non-existent namespace on a managed cluster, create it first

Signed-off-by: Travis Janssen <travis.janssen@ibm.com>

---
**Known issue**: the logic for verifying that a namespace exists on a managed cluster is housed in `drplacementcontrol.go::createVRGManifestWork()`, but if the namespace doesn't exist and none of the resources exist when Ramen initializes, `drplacementcontrol_controller.go::getVRGsFromManagedClusters()` executes first, which will result in Reconcile errors. However, this runs a few times then creates the namespace and those errors are resolved. I couldn't figure out how to generalize this functionality in a way that didn't require the DRPC instance or mwu_util instance and fixed the errors at start. On my test machine, this executed so quickly that I had to search for the errors, but it's probably worth fixing later.

--- 
**How to easily reproduce the error this patch fixes:**
1. Obtain a clean environment (meaning no minikube instances, no virsh osd pools), then run the following commands:
2. `$ bash ramen/hack/ocm-minikube-ramen.sh`
3. `$ bash ramen/hack/ocm-minikube-ramen.sh application_sample_deploy`
4. `$ kubectl delete namespace busybox-sample --context=cluster1`
5. `$ kubectl config use-context hub`
6. `$ make generate; make maniefsts; make; make install; make run-hub`

You should notice that since the `busybox-sample` namespace doesn't exist on cluster1, that VRG creation will fail.

ShyamsundarR commented on controllers/drplacementcontrol.go:
```diff
@@ -800,6 +814,29 @@ func (d *DRPCInstance) createVRGManifestWork(homeCluster string) error {
+		return fmt.Errorf("failed to verify if namespace '%s' on cluster %s exists: %w",
+			d.instance.Namespace, homeCluster, err)
+	}
+
```
(nit) check if namespaceExists first and return nil, avoids a bigger if block as in golang [practices](https://golang.org/doc/effective_go#if). Just a stylistic comment.

ShyamsundarR commented on controllers/drplacementcontrol.go:
```diff
@@ -727,9 +727,16 @@ func (d *DRPCInstance) cleanupSecondaries(skipCluster string) (bool, error) {
+			return false, err
+		}
+
+		mcvNameNS := BuildManagedClusterViewName(d.instance.Name, d.instance.Namespace, rmnutil.MWTypeNS)
```
Needs to be cleaned up during the call to `finalizeDRPC` as well.

ShyamsundarR replied:
@tjanssen3 This additional cleanup needs to be addressed.

ShyamsundarR reviewed (approved):
@BenamarMk The call to create NS is just pre VRG creation on the managed cluster, which is fine. We may loop out as we just created the namespace, but checking once if delaying this till point is fine.

BenamarMk commented on controllers/util/mw_util.go:
```diff
@@ -371,21 +398,33 @@ func (mwu *MWUtil) createOrUpdateManifestWork(
 		return fmt.Errorf("failed to delete ManifestWork for VRG in namespace %s (%w)", clusterName, err)
 	}
 
+	/*  The ManifestWork that created a Namespace is intentionally left on the server
```
Just leave this line and remove the rest of the comment and the commented code

BenamarMk commented on controllers/drplacementcontrol.go:
```diff
@@ -877,6 +914,35 @@ func (d *DRPCInstance) updateCondition(condition *metav1.Condition, idx int, rea
+	}
+
+	// sanity checks: only allow Active namespaces
+	if namespace == nil {
```
(nit) I would remove this extra code.  No reason to double test.  If no error, we just have to trust that we have a namespace object and if not, we will panic below anyway.

BenamarMk commented on controllers/drplacementcontrol.go:
```diff
@@ -785,6 +792,13 @@ func (d *DRPCInstance) updateUserPlacementRuleStatus(status plrv1.PlacementRuleS
 }
 
 func (d *DRPCInstance) createVRGManifestWork(homeCluster string) error {
+	err := d.ensureNamespaceExistsOnManagedCluster(homeCluster)
```
Thinking... should we check for the existence of the VRG MW here before we use the ManagedClusterView to retrieved the Namespace.  If MW exists, then we know that the namespace exists.  If it doesn't exist, then we don't know and hence, we make this call.  Again, just a thought, as MCV is very expensive and it is guaranteed it will not complete in one shot.

BenamarMk replied:
Another question, if let's say the MCV fails the first time getting the Namespace, with a failure other than NotFound.  The second time we get here, will it create a new MCV.  I doubt it.  That means there is a potential we get stuck here forever.  Is that also true for all the calls to getManagedClusterResource.  Hmm.... I just checked the code and I think we are in trouble here.  I think we should delete the MCV on any result so next time it will be recreated.  Again, just thinking.

BenamarMk reviewed (approved):
Minor comments, but I did have kind of a question/thought about using MCV to detect the namespace existence all the time. None should hold merging this PR though.

tjanssen3 commented:
Updated one more time based on feedback above (in particular, the `finalizeDRPC` comment).

I agree with @BenamarMk that checking VRG MW could be a substitute for a full MCV check on the Namespace, but did not add it to this PR - I added an issue (#298) and a TODO to make sure we don't forget about this. I believe the later comment about deleting the MCV after the namespace's existence is established can addressed at the same time. Otherwise, I think this should be good to go.
